R/derive_vars_trt_start.R

```diff
@@ -0,0 +1,152 @@
+#' data("ex")
+#' data("dm")
+#'
+#'dm %>%
+#'  derive_vars_trt_start(dataset_ex = ex,
+#'                        new_var=TRTSDTM,
+#'                        dtc=EXSTDTC,
+#'                        date_imputation = "first",
+#'                        time_imputation="first",
+#'                        min_dates = NULL,
+#'                        max_dates = NULL,
+#'                        order = vars(EXSTDTC,EXSEQ),
+#'                        subject_keys=vars(STUDYID,USUBJID),
+#'                        flag_imputation = "AUTO",
+#'                        filter_ex = EXDOSE>0 ,
+#'                        ord_filter="first") %>%
+#'  select(USUBJID, TRTSDTM)
+derive_vars_trt_start <- function(dataset,
+                                  dataset_ex,
+                                  filter_ex = NULL,
+                                  subject_keys,
+                                  new_var,
+                                  dtc,
+                                  date_imputation,
+                                  time_imputation,
+                                  min_dates = NULL,
+                                  max_dates = NULL,
+                                  flag_imputation = "AUTO",
+                                  order,
+                                  ord_filter) {
+
+  new_var <- assert_symbol(enquo(new_var))
+  dtc <- assert_symbol(enquo(dtc))
+  assert_data_frame(dataset, subject_keys)
+  assert_data_frame(dataset_ex, required_vars = quo_c(subject_keys, order))
+  assert_character_scalar(ord_filter, values = c("first", "last"), case_sensitive = F)
+
+  filter_ex <- assert_filter_cond(enquo(filter_ex), optional = TRUE)
+  filter_ex <- enquo(filter_ex)
+
+  assert_character_scalar(flag_imputation, values = c("auto", "both", "date", "time", "none"),
+                          case_sensitive = FALSE)
+
+  #Issue warning that new_var should be DTM
+  if (str_detect(rlang::quo_name(new_var), "DTM", negate = T)) {
+    msg <- sprintf(
+      "This function derives the treatment datetime, new_var must end in DTM.")
+    warn(msg)
+  }
+
+  add <- dataset_ex %>%
+    mutate(imputed_dtc = convert_dtc_to_dtm(!!dtc,
+                                            date_imputation = date_imputation,
+                                            time_imputation = time_imputation,
+                                            min_dates = min_dates,
+                                            max_dates = max_dates))
+
+  #filter EX data based on study requirements
+  if (quo_is_null(filter_ex)) {
+    add
+    msg <- sprintf(
+      "Input EX dataset has not been filtered. Check study requirement if ex needs to be filtered.")  # nolint
+    warn(msg)
+  } else {
+    add <- add %>% filter(!!filter_ex)
+  }
+
+  #pick last or first date in sequence
+  add <- add %>%
+    filter_extreme(order = order,
+                   by_vars = subject_keys,
+                   mode = ord_filter) %>%
+    mutate(!!new_var := imputed_dtc) %>%
```

Comment:
all the code below (up to line 149) should not be needed if `derive_vars_dtm()` is used.